stellargraph/core/utils.py

```diff
@@ -195,7 +195,7 @@ def preprocess_adj(adj, symmetric=True):
         if isinstance(k, int) and k >= 2:
             # default minimum degree for Chebyshev polynomials.
             T_k = chebyshev_polynomial(rescale_laplacian(normalized_laplacian(A)), k)
-            features = [features] + T_k
+            features = np.concatenate([features] + [t.todense() for t in T_k], axis=1)
```

Comment:
This seems like it's going to absolutely explode the memory usage of this, since the `todense` calls will create arrays with `(k + 1) * n * n` elements in memory. Maybe we could instead have everything be sparse as it was: that is, convert `features` to the appropriate sparse format and concatenate them sparsely?

What do you and @habiba-h  think?

Reply:
Yes, I agree,  we should convert `features` to sparse and then concatenate.

Reply:
I think there's going to be extra logic required downstream if `features` becomes a sparse matrix. For instance, there's currently a separate `FullBatchNodeSequence` vs `SparseFullBatchNodeSequence` that works with dense vs sparse adjacency matrices that get fed into keras. Does that sound right?

Reply:
I know a little bit about how and where sparse and dense adjacency matrices are used but not sure how the sparse features will be affected downstream.

Reply:
I agree, there should not be a transformation to dense matrices here. It was done because at the time there was no support for sparse matrices. You'll need to return sparse matrices here (not convert them to dense) and then convert a list of sparse matrices to dense in the `FullBatchNodeSequence` class or to index/value arrays in the `SparseFullBatchNodeSequence` class.

Reply:
One issue I've had for a long time with this is that the `GCN_Aadj_feats_op` accepts features & adjacency matrix A and returns "transformed" features & new matrix "A". However, what the Chebyshev function does is add the Chebyshev matrix polynomials of A to the list of features, which is wrong IMHO.

I think the function should return a list of matrices as A, and for the Chebyshev function this will be a the transformed list. As far as I know, none of the methods actually transform the features.

Note that the higher the polynomial order the denser these matrices will become.

Reply:
> I think the function should return a list of matrices as A, and for the Chebyshev function this will be a the transformed list. As far as I know, none of the methods actually transform the features.

> I think that this is what we need to do to support the Chebyshev method, where we change the functions to not take the features matrix and return a list of matrices (GCN, GAT etc will return a list of one).

Ah, thanks for the clarification! So if I understood correctly... this method can be something like `def transform_adj(adj: Adj, method="gcn", k=1) -> List[Adj]`, and I can update the `GCN` layer, `FullBatchNodeSequence` and `SparseFullBatchNodeSequence` to always use a list of adjacency matrices?

Reply:
Actually, after looking through the `rgcn` implementation a bit, I'm suspecting that plugging the chebyshev matrix polynomials of A into RGCN might basically have the desired effect? In which case, it might make sense to just combine the two classes `GCN` and `RGCN` instead of implementing the multi-adjacency-matrix-handling logic twice